chemprop/uncertainty/calibrator.py

```diff
@@ -79,57 +80,220 @@ def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
+    More detailes can be found in [1]_.
+
+    Parameters
+    ----------
+    alpha: float
+        Error rate, :math:`\alpha \in [0, 1]`
+
+    References
+    ----------
+    .. [1] Angelopoulos, A.N.; Bates, S.; "A Gentle Introduction to Conformal Prediction and Distribution-Free
+        Uncertainty Quantification." arXiv Preprint 2021, https://arxiv.org/abs/2107.07511
+    """
+
+    def __init__(self, alpha: float):
+        super().__init__()
+        self.alpha = torch.tensor(alpha)
+        self.qhats = torch.tensor([])
+
+    @staticmethod
+    def nonconformity_scores(preds: Tensor):
+        r"""Compute nonconformity score as the negative of the softmax output for the true class.
+
+        .. math::
+            s_i = -\hat{f}(X_i)_{Y_i}
+        """
+        return -preds
+
     def fit(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> None:
-        ...
-        return
+        scores = self.nonconformity_scores(uncs)
+        for j in range(uncs.shape[1]):
+            mask_j = mask[:, j]
+            targets_j = targets[:, j][mask_j]
+            scores_j = scores[:, j][mask_j]
+
+            scores_j = torch.gather(scores_j, 1, targets_j.unsqueeze(1)).squeeze(1)
+            num_data = targets_j.shape[0]
+            q_level = (torch.ceil((num_data + 1) * (1 - self.alpha)) / num_data).item()
+            qhat = torch.quantile(scores_j, q_level, interpolation="higher")
+            self.qhats = torch.cat((self.qhats, torch.tensor([qhat])))
 
     def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
-        ...
-        return
+        calibrated_preds = torch.zeros_like(uncs, dtype=torch.int)
+        scores = self.nonconformity_scores(uncs)
+
+        for j, qhat in enumerate(self.qhats):
+            calibrated_preds[:, j] = (scores[:, j] <= qhat).int()
+
+        return preds, calibrated_preds
 
 
 @UncertaintyCalibratorRegistry.register("conformal-adaptive")
-class ConformalAdaptiveMulticlassCalibrator(UncertaintyCalibrator):
-    def fit(self, preds: Tensor, uncs: Tensor, targets: Tensor, mask: Tensor) -> None:
-        ...
-        return
+class ConformalAdaptiveMulticlassCalibrator(ConformalMulticlassCalibrator):
+    @staticmethod
+    def nonconformity_scores(preds):
+        r"""Compute nonconformity score by greedily including classes in the classification set until it reach the true label.
 
-    def apply(self, preds: Tensor, uncs: Tensor) -> tuple[Tensor, Tensor]:
-        ...
-        return
+        .. math::
+            s(x, y) = \sum_{j=1}^{k} \hat{f}(x)_{\pi_j(x)}, \text{ where } y = \pi_k(x)
+
+        where :math:`\pi_k(x)` is the permutation of :math:`\{1,...,K\}` that sorts :math:`\hat{f}(X_{test})` from most likely to least likely.
+        """
+
+        sort_inds = torch.argsort(-preds, dim=2)
+        sorted_preds = torch.gather(preds, 2, sort_inds)
+        sorted_scores = sorted_preds.cumsum(dim=2)
+        unsort_inds = torch.argsort(sort_inds, dim=2)
```

Comment:
rather than re-establishing the original order of the sorted index, why don't you just keep track of the original order that was sorted?

```python
n, t, s = preds.shape
orig_index = torch.arange(s).repeat(n, t, 1)
unsorted_index = index_orig.gather(dim=2, index=sorted_index)
```

Reply:
The above code is copied from the v1 implementation. I am happy to improve it. The code you suggest here would make the `unsorted_index` be the same as `sorted_index`. I think what you want to do here is `torch.zeros_like(sorted_scores).scatter_(2, sorted_index, sorted_scores)`.

Reply:
I don't think we should just be copying code blindly. There's clearly some unnecessary work being done here because you're:
1. getting the index of sorted scores (L226)
2. using that index to sort the scores (L227)
3. recovering the original, unsorted index via _another_ sort (L229)

You shouldn't need to sort an index to recover the index of the original, unsorted list. You should be able to just store it and recover it because sorting is not a cheap operation.

Reply:
Agree! I have fixed it.